# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/anrajvarun9838/flyrank_internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

### My lane: Machine Learning — Search Ranking / Discoverability

I will study which measurable content and page-level signals are associated with stronger search visibility. I chose this lane because I want to use my machine learning skills to understand patterns in real search data and turn them into practical decision-support recommendations. Over the next 7 weeks, I will focus on measurable relationships in the provided dataset rather than trying to explain or predict Google's ranking system.

In [7]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
# Section 1 check
lane = "Machine Learning — Search Ranking / Discoverability"

print("Lane:", lane)
print("Focus: measurable relationships between page/content signals and search visibility.")

Lane: Machine Learning — Search Ranking / Discoverability
Focus: measurable relationships between page/content signals and search visibility.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

### Decision, action, and cost

**Decision:** Which measurable page or content signals should be investigated first when deciding how to improve search visibility?

**Who acts:** A content or SEO team could use the findings to prioritize which pages or signals deserve further investigation.

**Cost of a wrong call:** A wrong recommendation could lead to spending time and resources improving a signal that has little useful relationship with search visibility. It could also cause useful signals to be overlooked. Therefore, recommendations will be treated as decision-support rather than guaranteed ranking improvements.

**Research question:** Which measurable signals in the starter dataset are associated with stronger search visibility, and how large and consistent are those observed relationships?

In [8]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.

# Section 2 check
decision = "Prioritize measurable signals for further investigation."
actor = "Content / SEO team"
wrong_call_cost = "Time and resources may be spent on signals with little useful relationship to search visibility."

print("Decision:", decision)
print("Who acts:", actor)
print("Cost of wrong call:", wrong_call_cost)

Decision: Prioritize measurable signals for further investigation.
Who acts: Content / SEO team
Cost of wrong call: Time and resources may be spent on signals with little useful relationship to search visibility.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

### Quick look at the starter data

The starter dataset gives an initial view of the available search-related observations. I will use these real dataset values to check whether there is enough variation in the target and available signals to justify deeper analysis over the next 7 weeks.

The numbers below are descriptive only. They do not show that any particular signal causes better search visibility.

In [11]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.
import os
import pandas as pd
import subprocess
import sys

REPO_DIR = "/content/flyrank-ml-internship-starter"

# Clone the public starter repo if it is not already present
if not os.path.isdir(REPO_DIR):
    subprocess.run(
        [
            "git",
            "clone",
            "--depth",
            "1",
            "https://github.com/flyrank-bih/flyrank-ml-internship-starter",
            REPO_DIR,
        ],
        check=True,
    )

# Load the starter dataset
csv_path = os.path.join(
    REPO_DIR,
    "data",
    "raw",
    "content_refresh_anonymized.csv"
)

df = pd.read_csv(csv_path)

# Starter proxy label
declining = df["trend_direction"].str.lower().eq("down")

# Supporting numbers for the research question
total_pages = len(df)
declining_pages = declining.sum()
declining_rate = declining.mean()
visible_declining_pages = (
    declining & (df["impressions_90d"] >= 500)
).sum()

print(f"Total pages: {total_pages:,}")
print(f"Declining pages: {declining_pages:,}")
print(f"Declining rate: {declining_rate:.1%}")
print(f"Declining pages with >=500 impressions: {visible_declining_pages:,}")

Total pages: 30,000
Declining pages: 16,262
Declining rate: 54.2%
Declining pages with >=500 impressions: 9,961


##3. Quick look at the data (2-3 real numbers)
The starter dataset contains 30,000 content pages. Using the starter proxy label, 16,262 pages (54.2%) are classified as declining.

Among those declining pages, 9,961 have at least 500 impressions in the 90-day window. This suggests that there is a substantial set of visible pages where prioritizing human review could be useful.

These numbers make the refresh and content-opportunity problem worth investigating further, while the next stages will test whether better ranking methods can improve prioritization.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').

This work can identify observed relationships between content and search signals and help create a directional ranking of pages for human review. It can provide decision-support by showing which pages appear more worthy of attention based on the signals and validation methods used.

I cannot claim that the analysis proves why Google ranks a page in a certain way, that a recommended refresh will definitely improve performance, or that the model predicts Google's algorithm.

I also cannot treat the starter proxy label as proof of a future decline. Any future model should use clearly defined future outcomes and leakage-safe validation before stronger predictive claims are made.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [✔] Every section above is filled — markdown thinking AND the code that backs it
- [✔] The notebook runs top to bottom with no errors (Runtime → Run all)
- [✔] No client names, URLs, or private queries anywhere
- [✔] My claims use careful words: observed, measured, directional, decision-support
- [✔] Committed to my repo under work/notebooks/ — then submit your repo URL on the card. Done.